pkg/cmd/pulumi/util.go

```diff
@@ -839,3 +839,26 @@ func checkDeploymentVersionError(err error, stackName string) error {
 	}
 	return errors.Wrap(err, "could not deserialize deployment")
 }
+
+func getRefreshOption(proj *workspace.Project, refresh string) bool {
+	// we want to check for an explicit --refresh or a --refresh=true or --refresh=false
+	// no is the default we assign to the refresh default to be able to understand
+	// the difference between when the user actually interacted with the cli argument
+	// and the default functionality today
+	if refresh != "" && refresh != "no" {
```

Comment:
I'm concerned that we're not surfacing an error when faced with unexpected input, and that we're using a sort of random value (`"no"`) rather than a standard bool as the default here.

How about defining the `refresh` flag using [`VarP`](https://pkg.go.dev/github.com/spf13/pflag#FlagSet.Var), and defining an implementation of [`Value`](https://pkg.go.dev/github.com/spf13/pflag#Value) with "optional boolean" semantics? e.g.

```go
type optionalBoolean struct {
	value *bool
}

var _ = pflag.Value((*optionalBoolean)(nil))

func (b *optionalBoolean) String() string {
	if b.value == nil {
		return "<unset>"
	}
	return strconv.FormatBool(*b.value)
}

func (b *optionalBoolean) Set(s string) error {
	v, err := strconv.ParseBool(s)
	if err != nil {
		return err
	}
	b.value = &v
}

func (b *optionalBoolean) Type() string {
	return "bool"
}

...

func newFooCmd() {
	var refresh optionalBoolean
	var cmd = &cobra.Command{
		...
	}
	cmd.PersistentFlags().VarP(
 		&refresh, "refresh", "r",
 		"Refresh the state of the stack's resources before this update")
}

Reply:
@stack72 and I worked together to come up with a solution that essentially equates to the same thing. If you'd rather we can use the value `"<unset>"` instead of `no`. And yeah I agree it should return an error if we can't parse the value. But otherwise, these are just different ways of doing the same thing.

Reply:
And I kinda think the way it currently is is easier to read :)

Reply:
> And I kinda think the way it currently is is easier to read :)

I'll admit that I don't love the way the `flag`/`pflag` packages deal with user-defined parsing--I'd rather be able to just supply a function. The downside I see to the current approach is that it puts the error reporting and handling in an awkward place--at the point the value is calculated, rather than at the point the flag is parsed.

> If you'd rather we can use the value "<unset>" instead of "no"

If we stick with a plain string, is there any reason not to use `""` for the default value? I don't see a semantic difference between `""` and `"no"`, and `"no"` has the downside of being something of a one-off. FWIW, we could use "" in the `Value` approach as well.

Reply:
I didn't say `""` I said `"<unset>"` as in your example code

Reply:
I hear you and I respectfully think this is fine 🙏🏽

Reply:
> I didn't say "" I said "<unset>" as in your example code

Yeah, I know--I was making another suggestion--sorry for the lack of clarity!

Reply:
Yeah I think we arrived at `"no"` from somewhat of a meandering path and looking over it again it's true that `""` works just as well and is clearer and avoids the double `if refresh != "" && refresh != "no"`. 

Pretty sure we started with the empty string and I somehow convinced @stack72 that it wouldn't work. But it does!